src/Disks/HDFS/DiskHDFS.cpp

```diff
@@ -0,0 +1,235 @@
+        if (exist)
+            removeFile(path);
+
+        auto metadata = createMeta(path);
+        /// Save empty metadata to disk to have ability to get file size while buffer is not finalized.
+        metadata.save();
+
+        LOG_DEBUG(log,
+            "Write to file by path: {}. New hdfs path: {}", backQuote(metadata_path + path), hdfs_path);
+
+        return std::make_unique<WriteIndirectBufferFromHDFS>(config, hdfs_path, file_name, metadata, buf_size, O_WRONLY);
+    }
+    else
+    {
+        auto metadata = readMeta(path);
+
+        LOG_DEBUG(log,
+            "Append to file by path: {}. New hdfs path: {}. Existing HDFS objects: {}",
+            backQuote(metadata_path + path), hdfs_path, metadata.remote_fs_objects.size());
+
+        return std::make_unique<WriteIndirectBufferFromHDFS>(
+                config, hdfs_path, file_name, metadata, buf_size, O_WRONLY | O_APPEND);
+    }
+}
+
+
+void DiskHDFS::removeFromRemoteFS(const Metadata & metadata)
+{
+    for (const auto & [hdfs_object_path, _] : metadata.remote_fs_objects)
+    {
+        /// Add path from root to file name
+        const size_t begin_of_path = remote_fs_root_path.find('/', remote_fs_root_path.find("//") + 2);
+        const String hdfs_path = remote_fs_root_path.substr(begin_of_path) + hdfs_object_path;
+
+        int res = hdfsDelete(hdfs_fs.get(), hdfs_path.c_str(), 0);
+        if (res == -1)
+            throw Exception(ErrorCodes::LOGICAL_ERROR, "HDFSDelete failed with path: " + hdfs_path);
+    }
+}
+
+
+void DiskHDFS::removeSharedFile(const String & path, bool keep_in_remote_fs)
+{
+    removeMeta(path, keep_in_remote_fs);
+}
+
+
+void DiskHDFS::removeSharedRecursive(const String & path, bool keep_in_remote_fs)
+{
+    removeMetaRecursive(path, keep_in_remote_fs);
+}
+
+
+void DiskHDFS::removeFileIfExists(const String & path)
+{
+    if (Poco::File(metadata_path + path).exists())
+        removeMeta(path, /* keep_in_remote_fs */ false);
+}
+
+
+void DiskHDFS::removeRecursive(const String & path)
+{
+    checkStackSize(); /// This is needed to prevent stack overflow in case of cyclic symlinks.
+
+    Poco::File file(metadata_path + path);
+    if (file.isFile())
+    {
+        removeFile(path);
+    }
+    else
+    {
+        for (auto it{iterateDirectory(path)}; it->isValid(); it->next())
+            removeRecursive(it->path());
+        file.remove();
+    }
+}
+
+
+void DiskHDFS::removeMeta(const String & path, bool keep_in_remote_fs)
+{
+    Poco::File file(metadata_path + path);
+
+    if (!file.isFile())
+        throw Exception(ErrorCodes::CANNOT_DELETE_DIRECTORY, "Path '{}' is a directory", path);
+
+    try
+    {
+        auto metadata = readMeta(path);
+
+        /// If there is no references - delete content from remote FS.
+        if (metadata.ref_count == 0)
+        {
+            file.remove();
+
+            if (!keep_in_remote_fs)
+                removeFromRemoteFS(metadata);
```

Comment:
Looks like a copy-paste from DiskS3. Can we extract it to DiskRemote and make method `removeFromRemoteFS` as pure virtual with S3/HDFS implementation?

Reply:
I wanted to do so, but for `s3`, `removeRecursive` creates `AwsS3KeyKeeper` keys and recursively collects them while local meta is removed and only after they are removed from `s3` by a call to `removeAws(keys)`. And for `hdfs` each file in hdfs is removed once its meta in `ch` is removed - in `removeFromRemoteFS`. 

I mean this part:
```
void DiskS3::removeSharedRecursive(const String & path, bool keep_s3)
{
    AwsS3KeyKeeper keys;
    removeMetaRecursive(path, keys);
    if (!keep_s3)
        removeAws(keys);
}
```
Is it ok for s3 to also remove files once its meta is removed  without waiting to first collecting them all - as it is now? (of course, under same `keep_in_remote_fs` flag)